File: frontend/src/tipos/solicitud-asesor.ts
import type { Nullable } from './api';

// --- ENUMS (Coinciden con Backend Go) ---

export type EstadoSolicitud = 'PENDIENTE' | 'EN_ATENCION' | 'RESUELTO' | 'CANCELADO';
export type CanalOrigenSolicitud = 'WHATSAPP' | 'WEB' | 'TELEFONO';
export type PrioridadSolicitud = 'BAJA' | 'NORMAL' | 'ALTA' | 'URGENTE';

// --- INTERFACE PRINCIPAL ---

export interface SolicitudAsesor {
  id: string;
  tenant_id: string;

  nombre: string;
  telefono: string;
  motivo: string;

  canal_origen: CanalOrigenSolicitud;
  canal_whatsapp_id: Nullable<string>;

  estado: EstadoSolicitud;
  prioridad: PrioridadSolicitud;

  asignado_a: Nullable<string>;
  fecha_asignacion: Nullable<string>;
  fecha_resolucion: Nullable<string>;

  nota_interna: Nullable<string>;
  resumen_conversacion: Nullable<string>;

  fecha_creacion: string;
  fecha_actualizacion: string;

  // Campo calculado (JOIN con usuarios_admin)
  nombre_asesor?: string;
}

// --- DTOs (Payloads para enviar al Backend) ---

export interface CrearSolicitudAsesorRequest {
  nombre: string;
  telefono: string;
  motivo: string;
  canal_origen?: CanalOrigenSolicitud;
  canal_whatsapp_id?: string;
  prioridad?: PrioridadSolicitud;
  resumen_conversacion?: string;
}

export interface AsignarSolicitudRequest {
  asignado_a: string;
}

export interface ResolverSolicitudRequest {
  nota_interna?: string;
}

export interface ActualizarPrioridadRequest {
  prioridad: PrioridadSolicitud;
}

export interface ActualizarNotaInternaRequest {
  nota_interna: string;
}

// --- RESPONSE HELPERS ---

export interface ContadorPendientes {
  total: number;
}

// --- MENSAJES DE ATENCIÓN (Chat en vivo) ---

export type RemitenteMensaje = 'CLIENTE' | 'ASESOR' | 'SISTEMA';

export interface MensajeAtencion {
  tenant_id: string;
  id: string;
  solicitud_id: string;
  remitente: RemitenteMensaje;
  contenido: string;
  asesor_id: string | null;
  fecha_envio: string;
}

// Para el selector de reasignación
export interface UsuarioResumen {
  id: string;
  nombre_completo: string;
  email: string;
  rol: string;
}

File: backend/internal/model/solicitud_asesor.go
package model

import "time"

// SolicitudAsesor representa una solicitud de atención humana generada
// desde WhatsApp, web o teléfono. El flujo es:
// PENDIENTE → EN_ATENCION → RESUELTO | CANCELADO
type SolicitudAsesor struct {
	TenantModel

	Nombre   string `json:"nombre" db:"nombre"`
	Telefono string `json:"telefono" db:"telefono"`
	Motivo   string `json:"motivo" db:"motivo"`

	CanalOrigen     string   `json:"canal_origen" db:"canal_origen"`
	CanalWhatsAppID NullUUID `json:"canal_whatsapp_id" db:"canal_whatsapp_id"`

	Estado   string `json:"estado" db:"estado"`
	Prioridad string `json:"prioridad" db:"prioridad"`

	AsignadoA        NullUUID `json:"asignado_a" db:"asignado_a"`
	
	FechaAsignacion  NullTime `json:"fecha_asignacion" db:"fecha_asignacion"`
	FechaResolucion  NullTime `json:"fecha_resolucion" db:"fecha_resolucion"`

	NotaInterna          NullString `json:"nota_interna" db:"nota_interna"`
	ResumenConversacion  NullString `json:"resumen_conversacion" db:"resumen_conversacion"`

	FechaCreacion      time.Time `json:"fecha_creacion" db:"fecha_creacion"`
	FechaActualizacion time.Time `json:"fecha_actualizacion" db:"fecha_actualizacion"`

	// Campo calculado (LEFT JOIN con usuarios_admin)
	NombreAsesor NullString `json:"nombre_asesor" db:"nombre_asesor"`
}

// Estados de solicitud de asesor.
const (
	SolicitudPendiente  = "PENDIENTE"
	SolicitudEnAtencion = "EN_ATENCION"
	SolicitudResuelta   = "RESUELTO"
	SolicitudCancelada  = "CANCELADO"
)

// Canales de origen de solicitud.
const (
	CanalOrigenWhatsApp = "WHATSAPP"
	CanalOrigenWeb      = "WEB"
	CanalOrigenTelefono = "TELEFONO"
)

// Prioridades de solicitud.
const (
	PrioridadBaja    = "BAJA"
	PrioridadNormal  = "NORMAL"
	PrioridadAlta    = "ALTA"
	PrioridadUrgente = "URGENTE"
)

// EstaAbierta indica si la solicitud aún requiere atención.
func (s *SolicitudAsesor) EstaAbierta() bool {
	return s.Estado == SolicitudPendiente || s.Estado == SolicitudEnAtencion
}